src/nagini_translation/translators/permission.py

```diff
@@ -36,14 +38,113 @@ def translate_perm_Num(self, node: ast.Num, ctx: Context) -> Expr:
+                return self.viper.Mul(left, right,
+                                      self.to_position(node, ctx),
+                                      self.no_info(ctx))
+            if left_int or right_int:
+                return self.viper.IntPermMul(left, right,
                                              self.to_position(node, ctx),
                                              self.no_info(ctx))
-        raise UnsupportedException(node)
\ No newline at end of file
+
+            return self.viper.PermMul(left, right,
+                                      self.to_position(node, ctx),
+                                      self.no_info(ctx))
+
+        newnode = None
+        if isinstance(node.op, ast.Add):
+            newnode = self.viper.PermAdd
+        elif isinstance(node.op, ast.Sub):
+            newnode = self.viper.PermSub
+
+        if newnode:
+            left = self.translate_perm(node.left, ctx)
+            right = self.translate_perm(node.right, ctx)
+            return newnode(left, right,
+                           self.to_position(node, ctx),
+                           self.no_info(ctx))
+        raise UnsupportedException(node)
+
+    def translate_perm_Call(self, node: ast.Call, ctx: Context) -> Expr:
+        func_name = get_func_name(node)
+        if func_name == 'ARP':
+            if len(node.args) == 0:
+                return self.get_arp_for_context(node, ctx)
+            elif len(node.args) == 1:
+                arg0_stmt, arg0 = self.translate_expr(node.args[0], ctx, self.viper.Int)
+                if arg0_stmt:
+                    raise InvalidProgramException(node, 'purity.violated')
+                # arg = self.viper.IntLit(node.args[0].n, self.to_position(node, ctx), self.no_info(ctx))
+                formal_arg = self.viper.LocalVarDecl('count', self.viper.Int, self.to_position(node, ctx), self.no_info(ctx))
+                return self.viper.FuncApp('rdc', [arg0], self.to_position(node, ctx),
+                                          self.no_info(ctx), self.viper.Perm, [formal_arg])
+        elif func_name == 'getARP':
+            if len(node.args) == 1:
+                formal_arg = self.viper.LocalVarDecl('tk', self.viper.Ref, self.to_position(node, ctx), self.no_info(ctx))
+                arg0_stmt, arg0 = self.translate_expr(node.args[0], ctx, self.viper.Ref)
+                if arg0_stmt:
+                    raise InvalidProgramException(node, 'purity.violated')
+                return self.viper.FuncApp('rd_token', [arg0], self.to_position(node, ctx),
+                                          self.no_info(ctx), self.viper.Perm, [formal_arg])
+
+        call_stmt, call = self.translate_expr(node, ctx, self.viper.Int)
+        if not call_stmt:
+            return call
+        raise InvalidProgramException(node, 'purity.violated')
+
+    def translate_perm_Name(self, node: ast.Name, ctx: Context) -> Expr:
+        if node.id == 'RD_PRED':
```

Comment:
Add an else branch that raises an UnsupportedException (if there are other valid cases that just aren't supported at the moment) or InvalidProgramException (if there should never be a different case)